internal/tools/alloydbnla/alloydbnla.go

```diff
@@ -0,0 +1,169 @@
+	"strings"
+
+	"github.com/googleapis/genai-toolbox/internal/sources"
+	"github.com/googleapis/genai-toolbox/internal/sources/alloydbpg"
+	"github.com/googleapis/genai-toolbox/internal/tools"
+	"github.com/jackc/pgx/v5/pgxpool"
+)
+
+const ToolKind string = "alloydb-ai-nl"
+
+type compatibleSource interface {
+	PostgresPool() *pgxpool.Pool
+}
+
+// validate compatible sources are still compatible
+var _ compatibleSource = &alloydbpg.Source{}
+
+var compatibleSources = [...]string{alloydbpg.SourceKind}
+
+type Config struct {
+	Name                string          `yaml:"name" validate:"required"`
+	Kind                string          `yaml:"kind" validate:"required"`
+	Source              string          `yaml:"source" validate:"required"`
+	Description         string          `yaml:"description" validate:"required"`
+	NLConfig            string          `yaml:"nlConfig" validate:"required"`
+	AuthRequired       []string         `yaml:"authRequired"`
+	NLConfigParameters tools.Parameters `yaml:"nlConfigParameters"`
+}
+
+// validate interface
+var _ tools.ToolConfig = Config{}
+
+func (cfg Config) ToolConfigKind() string {
+	return ToolKind
+}
+
+func (cfg Config) Initialize(srcs map[string]sources.Source) (tools.Tool, error) {
+	// verify source exists
+	rawS, ok := srcs[cfg.Source]
+	if !ok {
+		return nil, fmt.Errorf("no source named %q configured", cfg.Source)
+	}
+
+	// verify the source is compatible
+	s, ok := rawS.(compatibleSource)
+	if !ok {
+		return nil, fmt.Errorf("invalid source for %q tool: source kind must be one of %q", ToolKind, compatibleSources)
+	}
+
+	paramNames := make([]string, 0, len(cfg.NLConfigParameters))
+	for _, paramDef := range cfg.NLConfigParameters {
+		paramNames = append(paramNames, paramDef.GetName())
+	}
+	quotedParamNames := make([]string, len(paramNames))
+	for i, name := range paramNames {
+		// Basic escaping for single quotes within the name itself
+		escapedName := strings.ReplaceAll(name, "'", "''")
+		quotedParamNames[i] = fmt.Sprintf("'%s'", escapedName)
+	}
+	paramNamesSQL := "ARRAY []" // Default for no parameters
+	if len(quotedParamNames) > 0 {
+		paramNamesSQL = fmt.Sprintf("ARRAY [%s]", strings.Join(quotedParamNames, ", "))
+	}
+	paramValuePlaceholders := make([]string, len(paramNames))
+	for i := 0; i < len(paramNames); i++ {
+		// Placeholders start from $2 ($1 is reserved for the natural language query)
+		paramValuePlaceholders[i] = fmt.Sprintf("$%d", i+2)
+	}
+	paramValuesSQL := "ARRAY []" // Default for no parameters
+	if len(paramValuePlaceholders) > 0 {
+		paramValuesSQL = fmt.Sprintf("ARRAY [%s]", strings.Join(paramValuePlaceholders, ", "))
+	}
+
+	// execute_nl_query is the AlloyDB AI function that executes the natural language query
+	// The first parameter is the natural language query, which is passed as $1
+	// The second parameter is the NLConfig, which is passed as a string
+	// The third and fourth parameters are the list of nl_config parameter names and values, respectively
```

Comment:
Base on above I don't think it's 3rd and 4th params, I think its "the following params".

It would also be nice to include an example of what the final SQL statement will look like to help cement.